python/private/bzlmod/multiarch_whl_library.bzl

```diff
@@ -0,0 +1,316 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This stores the `bzlmod` specific `multiarch_whl_library` that can make more
+assumptions about how the repositories are setup and we can also fetch some data
+from the PyPI via its simple API.
+
+There is a single `multiarch_whl_library` repository that only does aliases by
+target platform to the available whls and/or the whl built from sdist by bazel.
+The main algorithm is as below:
+1. Fetch data from all known PyPI indexes about where the wheels live.
+2. Ensure that we have found info about all of the packages by checking that we know
+   URL for each artifact that is mentioned by its sha256 in the requirements file.
+3. Infer the compatible platforms for the artifacts by parsing the last entry
+   in the URL. This allows us to create a set of `whl_library` repos that (mostly)
+   do not depend on the host platform.
+
+Shortcomings of the design:
+- We need to wait for the module extension to pull metadata about existing packgaes
+  for multiple distributions and it scales as 'number of different packgaes' x
+  'number of indexes'. We do some optimizations that make the scaling better,
+  but without `MODULE.bazel.lock` we need to do this every time if we haven't
+  cached the repositories.
```

Comment:
Almost, we need few more things:
* Remove the selection of the os specific requirements file in `pip.bzl` extension and just register all requirements.
* Introduce a `host` python repo rule that would allow us to have the same label for the python used in external repos no matter what.
* (long term) Ideally use the `DEFAULT_PYTHON` in all of the setup and extend the `target_platforms` wiring to also support the Python implementation as one of the fields (like `cp39`, `cpython`). This would allow us to better support any python version in general.

Reply:
For posterity, with #1644 the lock file contents would be the same on every platform if the only a single requirements lock file is used to create the hub repo even without this PR. However, I am not sure how to support both, multiple requirements files and cross-platformness without getting more information about what is in the lock file.